File: backend/core/performance.py
"""
Performans Optimizasyon Sistemi
Tüm performans iyileştirmeleri tek yerden yönetilir
"""

import time
import asyncio
import functools
import threading
from typing import Any, Callable, Dict, List, Optional, Union
from dataclasses import dataclass
from datetime import datetime, timedelta
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
from core.logging import get_logger, log_performance_metric
from core.config import settings

logger = get_logger('performance')

@dataclass
class PerformanceMetric:
    """Performans metriği"""
    name: str
    duration: float
    timestamp: datetime
    metadata: Dict[str, Any]
    success: bool
    error: Optional[str] = None

class PerformanceMonitor:
    """Performans izleyici"""
    
    def __init__(self):
        self.metrics: List[PerformanceMetric] = []
        self.max_metrics = 10000
        self.slow_threshold = 2.0  # seconds
        self.very_slow_threshold = 5.0  # seconds
        
    def record_metric(self, name: str, duration: float, metadata: Dict[str, Any] = None, success: bool = True, error: str = None):
        """Performans metriği kaydet"""
        metric = PerformanceMetric(
            name=name,
            duration=duration,
            timestamp=datetime.now(),
            metadata=metadata or {},
            success=success,
            error=error
        )
        
        self.metrics.append(metric)
        
        # Maintain max metrics
        if len(self.metrics) > self.max_metrics:
            self.metrics.pop(0)
        
        # Log slow operations
        if duration > self.very_slow_threshold:
            logger.warning(f"Very slow operation: {name} took {duration:.2f}s", extra=metadata)
        elif duration > self.slow_threshold:
            logger.info(f"Slow operation: {name} took {duration:.2f}s", extra=metadata)
        
        # Log performance metric
        log_performance_metric(name, duration, metadata=metadata)
    
    def get_metrics(self, name: Optional[str] = None, limit: int = 100) -> List[PerformanceMetric]:
        """Performans metriklerini al"""
        metrics = self.metrics
        
        if name:
            metrics = [m for m in metrics if m.name == name]
        
        return metrics[-limit:]
    
    def get_stats(self, name: Optional[str] = None) -> Dict[str, Any]:
        """Performans istatistiklerini al"""
        metrics = self.get_metrics(name)
        
        if not metrics:
            return {}
        
        durations = [m.duration for m in metrics]
        success_count = sum(1 for m in metrics if m.success)
        error_count = len(metrics) - success_count
        
        return {
            'count': len(metrics),
            'success_count': success_count,
            'error_count': error_count,
            'success_rate': success_count / len(metrics) if metrics else 0,
            'avg_duration': sum(durations) / len(durations),
            'min_duration': min(durations),
            'max_duration': max(durations),
            'total_duration': sum(durations),
            'slow_count': sum(1 for d in durations if d > self.slow_threshold),
            'very_slow_count': sum(1 for d in durations if d > self.very_slow_threshold),
        }
    
    def clear_metrics(self):
        """Metrikleri temizle"""
        self.metrics.clear()

# Global performance monitor
performance_monitor = PerformanceMonitor()

def measure_performance(name: Optional[str] = None, metadata: Dict[str, Any] = None):
    """Performans ölçüm dekoratörü"""
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            start_time = time.time()
            success = True
            error = None
            
            try:
                result = func(*args, **kwargs)
                return result
            except Exception as e:
                success = False
                error = str(e)
                raise
            finally:
                duration = time.time() - start_time
                metric_name = name or f"{func.__module__}.{func.__name__}"
                performance_monitor.record_metric(
                    metric_name, duration, metadata, success, error
                )
        
        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            start_time = time.time()
            success = True
            error = None
            
            try:
                result = await func(*args, **kwargs)
                return result
            except Exception as e:
                success = False
                error = str(e)
                raise
            finally:
                duration = time.time() - start_time
                metric_name = name or f"{func.__module__}.{func.__name__}"
                performance_monitor.record_metric(
                    metric_name, duration, metadata, success, error
                )
        
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator

class ConnectionPool:
    """Bağlantı havuzu yöneticisi"""
    
    def __init__(self, max_connections: int = 20):
        self.max_connections = max_connections
        self.connections: List[Any] = []
        self.lock = threading.Lock()
        
    def get_connection(self) -> Any:
        """Bağlantı al"""
        with self.lock:
            if self.connections:
                return self.connections.pop()
            return None
    
    def return_connection(self, connection: Any):
        """Bağlantıyı geri ver"""
        with self.lock:
            if len(self.connections) < self.max_connections:
                self.connections.append(connection)
    
    def close_all(self):
        """Tüm bağlantıları kapat"""
        with self.lock:
            for connection in self.connections:
                try:
                    connection.close()
                except:
                    pass
            self.connections.clear()

class AsyncTaskPool:
    """Asenkron görev havuzu"""
    
    def __init__(self, max_workers: int = 10):
        self.max_workers = max_workers
        self.executor = ThreadPoolExecutor(max_workers=max_workers)
        self.semaphore = asyncio.Semaphore(max_workers)
        
    async def submit(self, func: Callable, *args, **kwargs) -> Any:
        """Görev gönder"""
        async with self.semaphore:
            loop = asyncio.get_event_loop()
            return await loop.run_in_executor(self.executor, func, *args, **kwargs)
    
    async def submit_batch(self, tasks: List[tuple]) -> List[Any]:
        """Toplu görev gönder"""
        results = []
        for task in tasks:
            if len(task) == 2:
                func, args = task
                kwargs = {}
            else:
                func, args, kwargs = task
            
            result = await self.submit(func, *args, **kwargs)
            results.append(result)
        
        return results
    
    def shutdown(self):
        """Havuzu kapat"""
        self.executor.shutdown(wait=True)

class CacheOptimizer:
    """Cache optimizasyonu"""
    
    def __init__(self):
        self.cache_stats = {
            'hits': 0,
            'misses': 0,
            'evictions': 0,
            'size': 0,
        }
    
    def optimize_cache_key(self, key: str, prefix: str = "") -> str:
        """Cache anahtarını optimize et"""
        # Remove special characters and normalize
        normalized = key.lower().replace(' ', '_').replace('-', '_')
        normalized = ''.join(c for c in normalized if c.isalnum() or c == '_')
        
        if prefix:
            return f"{prefix}:{normalized}"
        return normalized
    
    def get_cache_stats(self) -> Dict[str, Any]:
        """Cache istatistiklerini al"""
        total_requests = self.cache_stats['hits'] + self.cache_stats['misses']
        hit_rate = self.cache_stats['hits'] / total_requests if total_requests > 0 else 0
        
        return {
            **self.cache_stats,
            'hit_rate': hit_rate,
            'miss_rate': 1 - hit_rate,
        }

class QueryOptimizer:
    """Sorgu optimizasyonu"""
    
    def __init__(self):
        self.query_stats = {}
        self.slow_queries = []
    
    def optimize_query(self, query: str) -> str:
        """Sorguyu optimize et"""
        # Remove extra whitespace
        query = ' '.join(query.split())
        
        # Add hints for common patterns
        if 'SELECT' in query.upper() and 'LIMIT' not in query.upper():
            # Add default limit for SELECT queries
            query += ' LIMIT 1000'
        
        return query
    
    def record_slow_query(self, query: str, duration: float, params: Dict[str, Any] = None):
        """Yavaş sorguyu kaydet"""
        self.slow_queries.append({
            'query': query,
            'duration': duration,
            'params': params or {},
            'timestamp': datetime.now(),
        })
        
        # Keep only last 100 slow queries
        if len(self.slow_queries) > 100:
            self.slow_queries.pop(0)
        
        logger.warning(f"Slow query detected: {duration:.2f}s", extra={
            'query': query,
            'params': params,
        })
    
    def get_slow_queries(self, limit: int = 50) -> List[Dict[str, Any]]:
        """Yavaş sorguları al"""
        return self.slow_queries[-limit:]

class MemoryOptimizer:
    """Bellek optimizasyonu"""
    
    def __init__(self):
        self.memory_stats = {
            'allocations': 0,
            'deallocations': 0,
            'peak_usage': 0,
            'current_usage': 0,
        }
    
    def get_memory_usage(self) -> Dict[str, Any]:
        """Bellek kullanımını al"""
        try:
            import psutil
            process = psutil.Process()
            memory_info = process.memory_info()
            
            return {
                'rss': memory_info.rss,  # Resident Set Size
                'vms': memory_info.vms,  # Virtual Memory Size
                'percent': process.memory_percent(),
                'available': psutil.virtual_memory().available,
                'total': psutil.virtual_memory().total,
            }
        except ImportError:
            return {'error': 'psutil not available'}
    
    def optimize_memory(self):
        """Belleği optimize et"""
        try:
            import gc
            gc.collect()
            logger.info("Memory optimization completed")
        except Exception as e:
            logger.error(f"Memory optimization failed: {e}")

class PerformanceProfiler:
    """Performans profiler"""
    
    def __init__(self):
        self.profiles = {}
        self.active_profiles = set()
    
    def start_profile(self, name: str):
        """Profil başlat"""
        self.active_profiles.add(name)
        self.profiles[name] = {
            'start_time': time.time(),
            'calls': 0,
            'total_time': 0,
            'min_time': float('inf'),
            'max_time': 0,
        }
    
    def end_profile(self, name: str):
        """Profil bitir"""
        if name in self.active_profiles:
            self.active_profiles.remove(name)
            
            if name in self.profiles:
                profile = self.profiles[name]
                duration = time.time() - profile['start_time']
                
                profile['calls'] += 1
                profile['total_time'] += duration
                profile['min_time'] = min(profile['min_time'], duration)
                profile['max_time'] = max(profile['max_time'], duration)
                profile['avg_time'] = profile['total_time'] / profile['calls']
    
    def get_profile_stats(self) -> Dict[str, Any]:
        """Profil istatistiklerini al"""
        return self.profiles.copy()

# Global instances
connection_pool = ConnectionPool()
async_task_pool = AsyncTaskPool()
cache_optimizer = CacheOptimizer()
query_optimizer = QueryOptimizer()
memory_optimizer = MemoryOptimizer()
performance_profiler = PerformanceProfiler()

def get_performance_stats() -> Dict[str, Any]:
    """Genel performans istatistiklerini al"""
    return {
        'performance_monitor': performance_monitor.get_stats(),
        'cache_optimizer': cache_optimizer.get_cache_stats(),
        'memory_optimizer': memory_optimizer.get_memory_usage(),
        'query_optimizer': {
            'slow_queries_count': len(query_optimizer.slow_queries),
            'recent_slow_queries': query_optimizer.get_slow_queries(10),
        },
        'performance_profiler': performance_profiler.get_profile_stats(),
    }

File: frontend/src/core/performance.ts
/**
 * Performans Optimizasyon Sistemi
 * Tüm performans iyileştirmeleri tek yerden yönetilir
 */

import { logger } from './logger';
import { eventBus } from './eventBus';

export interface PerformanceMetric {
  name: string;
  duration: number;
  timestamp: number;
  metadata: Record<string, any>;
  success: boolean;
  error?: string;
}

export interface PerformanceStats {
  count: number;
  successCount: number;
  errorCount: number;
  successRate: number;
  avgDuration: number;
  minDuration: number;
  maxDuration: number;
  totalDuration: number;
  slowCount: number;
  verySlowCount: number;
}

class PerformanceMonitor {
  private metrics: PerformanceMetric[] = [];
  private maxMetrics = 10000;
  private slowThreshold = 2000; // milliseconds
  private verySlowThreshold = 5000; // milliseconds

  recordMetric(
    name: string,
    duration: number,
    metadata: Record<string, any> = {},
    success: boolean = true,
    error?: string
  ): void {
    const metric: PerformanceMetric = {
      name,
      duration,
      timestamp: Date.now(),
      metadata,
      success,
      error,
    };

    this.metrics.push(metric);

    // Maintain max metrics
    if (this.metrics.length > this.maxMetrics) {
      this.metrics.shift();
    }

    // Log slow operations
    if (duration > this.verySlowThreshold) {
      logger.warn(`Very slow operation: ${name} took ${duration}ms`, metadata);
    } else if (duration > this.verySlowThreshold) {
      logger.info(`Slow operation: ${name} took ${duration}ms`, metadata);
    }

    // Emit performance event
    eventBus.emit('performance:metric', metric);
  }

  getMetrics(name?: string, limit: number = 100): PerformanceMetric[] {
    let metrics = this.metrics;

    if (name) {
      metrics = metrics.filter(m => m.name === name);
    }

    return metrics.slice(-limit);
  }

  getStats(name?: string): PerformanceStats {
    const metrics = this.getMetrics(name);

    if (metrics.length === 0) {
      return {
        count: 0,
        successCount: 0,
        errorCount: 0,
        successRate: 0,
        avgDuration: 0,
        minDuration: 0,
        maxDuration: 0,
        totalDuration: 0,
        slowCount: 0,
        verySlowCount: 0,
      };
    }

    const durations = metrics.map(m => m.duration);
    const successCount = metrics.filter(m => m.success).length;
    const errorCount = metrics.length - successCount;

    return {
      count: metrics.length,
      successCount,
      errorCount,
      successRate: successCount / metrics.length,
      avgDuration: durations.reduce((sum, d) => sum + d, 0) / durations.length,
      minDuration: Math.min(...durations),
      maxDuration: Math.max(...durations),
      totalDuration: durations.reduce((sum, d) => sum + d, 0),
      slowCount: durations.filter(d => d > this.slowThreshold).length,
      verySlowCount: durations.filter(d => d > this.verySlowThreshold).length,
    };
  }

  clearMetrics(): void {
    this.metrics = [];
  }
}

class ConnectionPool {
  private connections: any[] = [];
  private maxConnections: number;
  private lock = false;

  constructor(maxConnections: number = 20) {
    this.maxConnections = maxConnections;
  }

  async getConnection(): Promise<any> {
    while (this.lock) {
      await new Promise(resolve => setTimeout(resolve, 10));
    }

    this.lock = true;
    try {
      if (this.connections.length > 0) {
        return this.connections.pop();
      }
      return null;
    } finally {
      this.lock = false;
    }
  }

  returnConnection(connection: any): void {
    while (this.lock) {
      // Wait for lock
    }

    this.lock = true;
    try {
      if (this.connections.length < this.maxConnections) {
        this.connections.push(connection);
      }
    } finally {
      this.lock = false;
    }
  }

  closeAll(): void {
    this.connections.forEach(connection => {
      try {
        if (connection && typeof connection.close === 'function') {
          connection.close();
        }
      } catch (error) {
        logger.error('Error closing connection', { error });
      }
    });
    this.connections = [];
  }
}

class AsyncTaskPool {
  private maxWorkers: number;
  private activeTasks = 0;
  private queue: Array<() => Promise<any>> = [];

  constructor(maxWorkers: number = 10) {
    this.maxWorkers = maxWorkers;
  }

  async submit<T>(task: () => Promise<T>): Promise<T> {
    return new Promise((resolve, reject) => {
      this.queue.push(async () => {
        try {
          const result = await task();
          resolve(result);
        } catch (error) {
          reject(error);
        }
      });

      this.processQueue();
    });
  }

  private async processQueue(): Promise<void> {
    if (this.activeTasks >= this.maxWorkers || this.queue.length === 0) {
      return;
    }

    this.activeTasks++;
    const task = this.queue.shift()!;

    try {
      await task();
    } finally {
      this.activeTasks--;
      if (this.queue.length > 0) {
        this.processQueue();
      }
    }
  }

  async submitBatch<T>(tasks: Array<() => Promise<T>>): Promise<T[]> {
    const promises = tasks.map(task => this.submit(task));
    return Promise.all(promises);
  }
}

class CacheOptimizer {
  private cacheStats = {
    hits: 0,
    misses: 0,
    evictions: 0,
    size: 0,
  };

  optimizeCacheKey(key: string, prefix: string = ''): string {
    // Remove special characters and normalize
    const normalized = key.toLowerCase()
      .replace(/\s+/g, '_')
      .replace(/[^a-z0-9_]/g, '');

    return prefix ? `${prefix}:${normalized}` : normalized;
  }

  getCacheStats(): Record<string, any> {
    const totalRequests = this.cacheStats.hits + this.cacheStats.misses;
    const hitRate = totalRequests > 0 ? this.cacheStats.hits / totalRequests : 0;

    return {
      ...this.cacheStats,
      hitRate,
      missRate: 1 - hitRate,
    };
  }

  recordHit(): void {
    this.cacheStats.hits++;
  }

  recordMiss(): void {
    this.cacheStats.misses++;
  }

  recordEviction(): void {
    this.cacheStats.evictions++;
  }

  updateSize(size: number): void {
    this.cacheStats.size = size;
  }
}

class QueryOptimizer {
  private queryStats: Record<string, any> = {};
  private slowQueries: Array<{
    query: string;
    duration: number;
    params: Record<string, any>;
    timestamp: number;
  }> = [];

  optimizeQuery(query: string): string {
    // Remove extra whitespace
    query = query.replace(/\s+/g, ' ').trim();

    // Add hints for common patterns
    if (query.toUpperCase().includes('SELECT') && !query.toUpperCase().includes('LIMIT')) {
      query += ' LIMIT 1000';
    }

    return query;
  }

  recordSlowQuery(
    query: string,
    duration: number,
    params: Record<string, any> = {}
  ): void {
    this.slowQueries.push({
      query,
      duration,
      params,
      timestamp: Date.now(),
    });

    // Keep only last 100 slow queries
    if (this.slowQueries.length > 100) {
      this.slowQueries.shift();
    }

    logger.warn('Slow query detected', { query, duration, params });
  }

  getSlowQueries(limit: number = 50): Array<Record<string, any>> {
    return this.slowQueries.slice(-limit);
  }
}

class MemoryOptimizer {
  private memoryStats = {
    allocations: 0,
    deallocations: 0,
    peakUsage: 0,
    currentUsage: 0,
  };

  getMemoryUsage(): Record<string, any> {
    if ('memory' in performance) {
      const memory = (performance as any).memory;
      return {
        usedJSHeapSize: memory.usedJSHeapSize,
        totalJSHeapSize: memory.totalJSHeapSize,
        jsHeapSizeLimit: memory.jsHeapSizeLimit,
        allocationCount: memory.allocationCount || 0,
      };
    }

    return { error: 'Memory API not available' };
  }

  optimizeMemory(): void {
    // Force garbage collection if available
    if ('gc' in window && typeof (window as any).gc === 'function') {
      (window as any).gc();
      logger.info('Memory optimization completed');
    } else {
      logger.info('Memory optimization not available');
    }
  }
}

class PerformanceProfiler {
  private profiles: Record<string, {
    startTime: number;
    calls: number;
    totalTime: number;
    minTime: number;
    maxTime: number;
    avgTime: number;
  }> = {};
  private activeProfiles = new Set<string>();

  startProfile(name: string): void {
    this.activeProfiles.add(name);
    this.profiles[name] = {
      startTime: performance.now(),
      calls: 0,
      totalTime: 0,
      minTime: Infinity,
      maxTime: 0,
      avgTime: 0,
    };
  }

  endProfile(name: string): void {
    if (this.activeProfiles.has(name)) {
      this.activeProfiles.delete(name);

      if (this.profiles[name]) {
        const profile = this.profiles[name];
        const duration = performance.now() - profile.startTime;

        profile.calls++;
        profile.totalTime += duration;
        profile.minTime = Math.min(profile.minTime, duration);
        profile.maxTime = Math.max(profile.maxTime, duration);
        profile.avgTime = profile.totalTime / profile.calls;
      }
    }
  }

  getProfileStats(): Record<string, any> {
    return { ...this.profiles };
  }
}

class ImageOptimizer {
  private imageCache = new Map<string, HTMLImageElement>();
  private maxCacheSize = 100;

  async loadOptimizedImage(src: string, options: {
    width?: number;
    height?: number;
    quality?: number;
  } = {}): Promise<HTMLImageElement> {
    const cacheKey = `${src}_${options.width || 'auto'}_${options.height || 'auto'}`;

    if (this.imageCache.has(cacheKey)) {
      return this.imageCache.get(cacheKey)!;
    }

    return new Promise((resolve, reject) => {
      const img = new Image();
      
      img.onload = () => {
        // Add to cache
        if (this.imageCache.size >= this.maxCacheSize) {
          const firstKey = this.imageCache.keys().next().value;
          this.imageCache.delete(firstKey);
        }
        this.imageCache.set(cacheKey, img);
        
        resolve(img);
      };
      
      img.onerror = () => {
        reject(new Error(`Failed to load image: ${src}`));
      };
      
      img.src = src;
    });
  }

  preloadImages(urls: string[]): Promise<HTMLImageElement[]> {
    const promises = urls.map(url => this.loadOptimizedImage(url));
    return Promise.all(promises);
  }

  clearCache(): void {
    this.imageCache.clear();
  }
}

class BundleOptimizer {
  private loadedChunks = new Set<string>();
  private chunkCache = new Map<string, any>();

  async loadChunk(chunkName: string): Promise<any> {
    if (this.chunkCache.has(chunkName)) {
      return this.chunkCache.get(chunkName);
    }

    try {
      const chunk = await import(/* webpackChunkName: "[request]" */ `../chunks/${chunkName}`);
      this.chunkCache.set(chunkName, chunk);
      this.loadedChunks.add(chunkName);
      return chunk;
    } catch (error) {
      logger.error('Failed to load chunk', { chunkName, error });
      throw error;
    }
  }

  preloadChunks(chunkNames: string[]): Promise<any[]> {
    const promises = chunkNames.map(name => this.loadChunk(name));
    return Promise.all(promises);
  }

  getLoadedChunks(): string[] {
    return Array.from(this.loadedChunks);
  }
}

// Global instances
export const performanceMonitor = new PerformanceMonitor();
export const connectionPool = new ConnectionPool();
export const asyncTaskPool = new AsyncTaskPool();
export const cacheOptimizer = new CacheOptimizer();
export const queryOptimizer = new QueryOptimizer();
export const memoryOptimizer = new MemoryOptimizer();
export const performanceProfiler = new PerformanceProfiler();
export const imageOptimizer = new ImageOptimizer();
export const bundleOptimizer = new BundleOptimizer();

// Performance measurement decorator
export function measurePerformance(
  name?: string,
  metadata: Record<string, any> = {}
) {
  return function (target: any, propertyName: string, descriptor: PropertyDescriptor) {
    const method = descriptor.value;

    descriptor.value = async function (...args: any[]) {
      const startTime = performance.now();
      const metricName = name || `${target.constructor.name}.${propertyName}`;
      let success = true;
      let error: string | undefined;

      try {
        const result = await method.apply(this, args);
        return result;
      } catch (err) {
        success = false;
        error = err instanceof Error ? err.message : String(err);
        throw err;
      } finally {
        const duration = performance.now() - startTime;
        performanceMonitor.recordMetric(metricName, duration, metadata, success, error);
      }
    };

    return descriptor;
  };
}

// Performance measurement function
export function measureFunction<T extends (...args: any[]) => any>(
  fn: T,
  name?: string,
  metadata: Record<string, any> = {}
): T {
  return (async (...args: any[]) => {
    const startTime = performance.now();
    const metricName = name || fn.name || 'anonymous';
    let success = true;
    let error: string | undefined;

    try {
      const result = await fn(...args);
      return result;
    } catch (err) {
      success = false;
      error = err instanceof Error ? err.message : String(err);
      throw err;
    } finally {
      const duration = performance.now() - startTime;
      performanceMonitor.recordMetric(metricName, duration, metadata, success, error);
    }
  }) as T;
}

export function getPerformanceStats(): Record<string, any> {
  return {
    performanceMonitor: performanceMonitor.getStats(),
    cacheOptimizer: cacheOptimizer.getCacheStats(),
    memoryOptimizer: memoryOptimizer.getMemoryUsage(),
    queryOptimizer: {
      slowQueriesCount: queryOptimizer.getSlowQueries().length,
      recentSlowQueries: queryOptimizer.getSlowQueries(10),
    },
    performanceProfiler: performanceProfiler.getProfileStats(),
    imageOptimizer: {
      cacheSize: (imageOptimizer as any).imageCache.size,
    },
    bundleOptimizer: {
      loadedChunks: bundleOptimizer.getLoadedChunks(),
    },
  };
}

export default {
  performanceMonitor,
  connectionPool,
  asyncTaskPool,
  cacheOptimizer,
  queryOptimizer,
  memoryOptimizer,
  performanceProfiler,
  imageOptimizer,
  bundleOptimizer,
  measurePerformance,
  measureFunction,
  getPerformanceStats,
};
